src/celeritas/ext/detail/HitProcessor.cc

```diff
@@ -380,14 +380,20 @@ void HitProcessor::operator()(DetectorStepOutput const& out, size_type i) const
  *
  * This is a bit like \c G4Step::UpdateTrack .
  */
-void HitProcessor::update_track(ParticleId id) const
+void HitProcessor::update_track(DetectorStepOutput const& out, size_type i) const
 {
+    ParticleId id = out.particle[i];
     CELER_EXPECT(id < tracks_.size());
     G4Track& track = *tracks_[id.unchecked_get()];
     step_->SetTrack(&track);
 
     G4ParticleDefinition const& pd = *track.GetParticleDefinition();
 
+    if (!out.parent_id.empty())
+    {
+        track.SetTrackID(out.parent_id[i].unchecked_get());
```

Comment:
I suppose there would be an issue there when reconstructing hit from a secondary, in which case the parent id points to Celeritas' TrackID

Reply:
Right, as part of #1810 we want to preserve the original Geant4 TrackID of offloaded primaries when reconstructing the track. I suppose it would be fine if internally Celeritas still uses its own primary ID as long as we can map it back to the Geant4 TrackID